### PR TITLE
Wire offline routing for ipns in core.Resolve

### PR DESCRIPTION
**Pull request by rht**
With this commit, ipfs get, ls, refs, tar, object can resolve ipns
entries from local cache when offline.

### CODE REVIEW

File: core/pathresolver.go [hunk 1]
```diff
@@ -23,9 +17,10 @@ func Resolve(ctx context.Context, n *IpfsNode, p path.Path) (*merkledag.Node, er
 	if strings.HasPrefix(p.String(), "/ipns/") {
 		// resolve ipns paths
 
-		// TODO(cryptix): we sould be able to query the local cache for the path
-		if n.Namesys == nil {
-			return nil, ErrNoNamesys
+		if n.Namesys == nil && !n.OnlineMode() {
+			if err := n.SetupOfflineRouting(); err != nil {
+				return nil, err
+			}
```
Comment (jbenet) [inline core/pathresolver.go]: no, this is not a good interface UX. very magical.

**Comment by jbenet** [comment]
@rht thanks, but this is not a good interface UX change.

**Comment by rht** [comment]
I can't see the situation when the daemon is off and the keys cache don't need be loaded when resolving ipns.

To resolve offline ipns for the commands above (get, ls, refs, tar, object), then `n.SetupOfflineRouting` will have to be added manually in each of them, which reason not entirely known to the reader. At least not until later, where in this case it is because it is for `core.Resolve`, or in other case whenever an ipns operation is present.

**Comment by jbenet** [comment]
i understand, but it's better than magically inserting it into resolve. this can be rethought as we refine teh core interfaces, prob for 0.5.0

**Comment by rht** [comment]
Does the 'magic' here refer to the side effect of loading the local cache when the functionality of `core.Resolve` is only to code-select between ipfs and ipns resolver? If this is the case, what about hiding the cache-loading in `Namesys.Resolve`?

Or does it refer to the fact of auto-loading the local cache itself? But there is no other way to do the cache-loading *on demand*.

**Comment by jbenet** [comment]
the "magic" is "hiding". it should not be hidden under unrelated APIs, it should be explicit. side effects are really problematic in code, and this is one example that would lead to problems down the road.

**Comment by rht** [comment]
So for now I should accept the fact that for every ipns operation in the code, the node must be initialized with the `SetupOfflineRouting`?

**Comment by jbenet** [comment]
yes. or trying to work it into core or resolver. i do not think this is worth much time right now

**Pull request closed without merging**